core/API/DataTableManipulator/ReportTotalsCalculator.php

```diff
@@ -83,30 +84,54 @@ protected function manipulateDataTable($dataTable)
             return $dataTable;
         }
 
-        $this->totals       = array();
-        $firstLevelTable    = $this->makeSureToWorkOnFirstLevelDataTable($dataTable);
-        $metricsToCalculate = Metrics::getMetricIdsToProcessReportTotal();
+        if (isset($this->request['totals']) && ($this->request['totals'] === 0 || $this->request['totals'] === '0')) {
+            return $dataTable;
+        }
 
-        $metricNames = array();
-        foreach ($metricsToCalculate as $metricId) {
-            $metricNames[$metricId] = Metrics::getReadableColumnName($metricId);
+        if ($dataTable instanceof DataTable\Simple) {
+            return $dataTable;
         }
 
-        if (!empty($this->report)) {
-            $reportMetrics = $this->report->getMetricNamesToProcessReportTotals();
-            foreach ($reportMetrics as $metricId => $metricName) {
-                $metricNames[$metricId] = $metricName;
+        $this->totals    = array();
+        $firstLevelTable = $this->makeSureToWorkOnFirstLevelDataTable($dataTable);
+      //  $firstLevelTable->applyQueuedFilters();
+
+        $clone = $firstLevelTable->getEmptyClone();
+
+        /** @var DataTable\Row $totalRow */
+        $totalRow = null;
+        foreach ($firstLevelTable->getRows() as $row) {
+            if (!isset($totalRow)) {
+                $columns = array('label' => 'Total12345') + $row->getColumns();
+                $totalRow = new DataTable\Row(array(DataTable\Row::COLUMNS => $columns));
+            } else {
+                $totalRow->sumRow(
+                    $row, $enableCopyMetadata = false, $firstLevelTable->getMetadata(DataTable::COLUMN_AGGREGATION_OPS_METADATA_NAME));
             }
         }
+        $clone->addRow($totalRow);
 
-        foreach ($firstLevelTable->getRows() as $row) {
-            $columns = $row->getColumns();
-            foreach ($metricNames as $metricId => $metricName) {
-                $this->sumColumnValueToTotal($columns, $metricId, $metricName);
+        if (array_keys($this->report->getProcessedMetrics()) === array('nb_actions_per_visit', 'avg_time_on_site', 'bounce_rate', 'conversion_rate')) {
+            // hack for AllColumns table
+            $clone->filter('AddColumnsProcessedMetrics');
+        }
+
+        $request = $this->request;
+        $request['totals'] = 0;
+        $request['disable_generic_filters'] = 1;
+
+        $processor = new DataTablePostProcessor($this->report->getModule(), $this->report->getAction(), $request);
+        $clone = $processor->process($clone);
+
+        foreach ($clone->getRows() as $row) {
+            if ($row->getColumn('label') === 'Total12345') {
+
+                $this->totals = $row->getColumns();
+                $dataTable->setMetadata('totals', $this->totals);
```

Comment:
We could likely also set here something like `$dataTable->setTotalsRow()` so plugins can at least have the chance to hook onto it... 

It's not that easy though. Tried eg this:

```diff
diff --git a/core/API/DataTableManipulator/ReportTotalsCalculator.php b/core/API/DataTableManipulator/ReportTotalsCalculator.php
index 4baac2f939..b1439f61ae 100644
--- a/core/API/DataTableManipulator/ReportTotalsCalculator.php
+++ b/core/API/DataTableManipulator/ReportTotalsCalculator.php
@@ -127,6 +127,8 @@ class ReportTotalsCalculator extends DataTableManipulator
             if ($row->getColumn('label') === 'Total12345') {
 
                 $this->totals = $row->getColumns();
+                $row->setColumn('label', 'Totals');
+                $dataTable->setTotalsRow($totalRow);
                 $dataTable->setMetadata('totals', $this->totals);
                 break;
             }
diff --git a/core/DataTable.php b/core/DataTable.php
index 22b09b97ee..9189170dac 100644
--- a/core/DataTable.php
+++ b/core/DataTable.php
@@ -301,6 +301,11 @@ class DataTable implements DataTableInterface, \IteratorAggregate, \ArrayAccess
      */
     protected $summaryRow = null;
 
+    /**
+     * @var \Piwik\DataTable\Row
+     */
+    protected $totalsRow = null;
+
     /**
      * Table metadata. Read [this](#class-desc-the-basics) to learn more.
      *
@@ -402,6 +407,16 @@ class DataTable implements DataTableInterface, \IteratorAggregate, \ArrayAccess
         }
     }
 
+    public function setTotalsRow(Row $totals)
+    {
+        $this->totalsRow = $totals;
+    }
+
+    public function getTotalsRow()
+    {
+        return $this->totalsRow;
+    }
+
     /**
      * Returns the name of the column this table was sorted by (if any).
      *
diff --git a/core/Plugin/Visualization.php b/core/Plugin/Visualization.php
index 614aefcffa..82129b8f8b 100644
--- a/core/Plugin/Visualization.php
+++ b/core/Plugin/Visualization.php
@@ -209,9 +209,8 @@ class Visualization extends ViewDataTable
         }
 
         if ($this->dataTable && $this->dataTable instanceof DataTable) {
-            $totals = $this->dataTable->getMetadata('totals');
+            $totals = $this->dataTable->getTotalsRow();
             if (!empty($totals)) {
-                $totals['label'] = 'Totals';
                 $this->dataTable->addRow(new DataTable\Row(array(DataTable\Row::COLUMNS => $totals)));
             }
         }
```
but breaks eg totals feature when flattening as it seems to create an empty clone currently and not copy totals row etc (to be implemented)